### PR TITLE
math: move to es6 classes

### PR DESCRIPTION
**Pull request by yomotsu**
A part of https://github.com/mrdoob/three.js/issues/19986

This PR contains:

- [x] Vector2
- [x] Vector3
- [x] Vector4
- [x] Matrix3
- [x] Matrix4
- [x] Quaternion
- [x] Euler

The rest of the math folder will be continued in separate PR.

---

The following are already completed in other PRs :

- Ray ( https://github.com/mrdoob/three.js/pull/19980 )
- Spherical ( https://github.com/mrdoob/three.js/pull/19935 )

### CODE REVIEW

File: src/math/Vector3.js [hunk 1]
```diff
@@ -1,22 +1,26 @@
 import { MathUtils } from './MathUtils.js';
 import { Quaternion } from './Quaternion.js';
 
-const _vector = new Vector3();
+let _vector;
```
Comment (yomotsu) [inline src/math/Vector3.js:4]: Hoisting does not work anymore, because Vector3 is not a function.
Vec3 instance will be assigned at the end of this file.

**Comment by yomotsu** [comment]
I have converted the following classes in the math folder so far:

- Vector2
- Vector3
- Vector4
- Matrix3
- Matrix4
- Quaternion
- Euler

@mrdoob @Mugen87 
Would you mind reviewing them, please? 🙇‍♂️

If these look fine, I will continue the rest of the math folder on other PR.

File: src/math/Vector3.js [hunk 2]
```diff
@@ -721,7 +725,8 @@ Object.assign( Vector3.prototype, {
 
 	}
 
-} );
+}
 
+_vector = new Vector3();
```
Comment (mrdoob) [inline src/math/Vector3.js:730]: Would it work to do `const _vector = new Vector3();` here?

Comment (yomotsu) [reply src/math/Vector3.js:730]: That works! (I tested)
So, other internal use instances should be declared at the bottom of the file too?

Such as: `const _quaternion = new Quaternion();` in
https://github.com/mrdoob/three.js/blob/f549b793fe070b951fa1018af8cbaddf8650c11f/src/math/Vector3.js#L1-L7
and
https://github.com/mrdoob/three.js/blob/f549b793fe070b951fa1018af8cbaddf8650c11f/src/math/Matrix4.js#L1-L11
also maybe
https://github.com/mrdoob/three.js/blob/f549b793fe070b951fa1018af8cbaddf8650c11f/src/math/Euler.js#L4-L9

Comment (mrdoob) [reply src/math/Vector3.js:730]: > So, other internal use instances should be declared at the bottom of the file too?

Yes, I think so!

Comment (yomotsu) [reply src/math/Vector3.js]: Okay, done it with 8bcfccd

File: src/math/Euler.js [hunk 3]
```diff
@@ -3,99 +3,88 @@ import { Vector3 } from './Vector3.js';
...
 
-		},
+		return [ 'XYZ', 'YZX', 'ZXY', 'XZY', 'YXZ', 'ZYX' ];
 
-		set: function ( value ) {
+	}
 
-			this._x = value;
-			this._onChangeCallback();
+	static get DefaultOrder() {
```
Comment (mrdoob) [inline src/math/Euler.js:23]: We should do a setter for this one too.

Comment (yomotsu) [reply src/math/Euler.js:23]: @mrdoob 

Thank you for reviewing.
Sorry, I got it wrong, I thought `DefaultOrder` is read-only too.
I will fix it.

Before that, I have a question: How do we declare Public class fields?

The [readable-writable static value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields) can be declare like
```js
class MyClass {

  static xxx = 1;

  constructor( x ) {
    this.x = x
  }

}
```

However It seems Buble transpiler [does not support the syntax](https://buble.surge.sh/#class%20MyClass%20%7B%0A%20%20%0A%20%20static%20xxx%20%3D%201%3B%0A%0A%20%20constructor(%20x%20)%20%7B%0A%20%20%20%20this.x%20%3D%20x%0A%20%20%7D%0A%0A%7D), while [Babel support that](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&spec=false&loose=false&code_lz=MYGwhgzhAECyCeBhcVoG8BQ1pehALmPgJbDRgXQC80AjANwa7AD2AdgQE4Cuw-LnABTQAHtACU6XNnwALYhAB0YmiNwBfJuqA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Ces2015%2Cstage-0%2Cenv&prettier=false&targets=&version=7.11.1&externalPlugins=)

Which option would be preferable in three.js (with Buble)?

```js
class MyClass {

  constructor( x ) {
    this.x = x
  }

}

MyClass.xxx = 1;
```
or
```js
let _xxx = 1;

class MyClass {

  static get xxx = () => {
    return _xxx;
  }

  static set xxx = ( value ) => {
    _xxx = value;
  }

  constructor( x ) {
    this.x = x
  }

}
```

or other?
Thank you in advance🙇‍♂️

maybe @DefinitelyMaybe or @ianpurvis could give a suggestion?

**Comment by DefinitelyMaybe** [comment]
[static is a keyword for methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/static) last time I checked. will have a look around.

File: src/math/Euler.js (hunk 3 above)
Comment (ianpurvis) [reply src/math/Euler.js:23]: > ```js
> class MyClass {
> 
>   constructor( x ) {
>     this.x = x
>   }
> 
> }
> 
> MyClass.xxx = 1;
> ```
 I think this style is nice for read/write 👍

Comment (yomotsu) [reply src/math/Euler.js]: Thanks, I'll do that then!

**Comment by yomotsu** [comment]
Fixed `Euler.DefaultOrder` to be readable-writable🙇‍♂️

File: src/math/Euler.js (hunk 3 above)
Comment (DefinitelyMaybe) [reply src/math/Euler.js]: For future reference:
There are only three things in the current `r120` dev build where Buble is being used.
`Curves`, `Geometries` and `Materials`.
When Buble is no longer needed, i.e. after converting everything to es6 classes, we should have a look back into this
```javascript
class MyClass {

  static xxx = 1;

  constructor( x ) {
    this.x = x
  }

}
```

Comment (DefinitelyMaybe) [reply src/math/Euler.js]: nvm. I've got Buble's purpose backwards. Buble is our 'convert classes back to support IE' tool and output it as `build/three.js`.

**Comment by mrdoob** [comment]
Can you also do `Euler.RotationOrders = [ 'XYZ', 'YZX', 'ZXY', 'XZY', 'YXZ', 'ZYX' ];`?

**Comment by mrdoob** [comment]
Should we also do `Euler.isEuler = true;`?

**Comment by yomotsu** [comment]
I'll revert `Euler.RotationOrders`.

However, I personally think `is***` properties are okay to be getters, such as `Euler.isEuler`.

- `is***` properties have been removed from the document at https://github.com/mrdoob/three.js/pull/18217.
- Users could change `is***` however not recommended in the old document https://github.com/mrdoob/three.js/commit/8badf776ea0c098c14c6359b0fb38e5d4918c6fc#diff-636a6ba5832420da16cd6cd99236e56a
- `is***` are read-only according to the d.ts https://github.com/mrdoob/three.js/blob/dev/src/math/Euler.d.ts#L13

Sould we do `Euler.isEuler = true;`?

**Comment by ianpurvis** [comment]
I think we should give the read-only getter a try? That feels like the right spirit for type checking. I'm not a benchmark master, but I ran my [cheap benchmark](https://hmyb.glitch.me/#MYGwhgzhAEBiD29oG8BQ1rHgOwgFwCcBXYPeAgCgEoV0No8ALASwgDpnc8xtgBTAAoF4AB2gBeBsT4BuOgF86Acz55oKvHj6UaaetAKqiBbFKKyFqRQnhsRwsngCeIvnYfxnroaIlmLqKhYXNAAZoh+2HwA7nCI1HKoAEZ8vIwUAESGYAAmALQ4IE7qqloEGQA00NQSAHxhiGwaZVSJKWmZ9p6eLnzQXa4EzpXVNOL14bZdjr0+Iq2B7cDpGZz4PPz9woPDVTXjDbZr3LyC2wvJqcuZxxt88KEj+xMRt6cPcfCtQA) on some different type checking. TLDR, performance is all similar.

Vanilla `instanceof` seems fast these days. That has got to be the nicest semantically, might be a good direction for after all the ES6 class changes stabilize.

**Comment by mrdoob** [comment]
> I think we should give the read-only getter a try?

I don't think this is the right time to give things a try. I think we should focus on porting the code to es6 modules with the least side effects as possible.

> Vanilla `instanceof` seems fast these days

We used to use `instanceof`. We changed it to `is***` to allow tree-shaking.

**Comment by mrdoob** [comment]
> However, I personally think `is***` properties can be getters, such as `Euler.isEuler`.

It doesn't feel good to turn `isEuler` into a function. These variables get accessed a lot per frame.

**Comment by ianpurvis** [comment]
Fair enough, let's go with `Euler.isEuler = true;` 👍

**Comment by yomotsu** [comment]
I understand performance is paramount.
I changed is** to be a prototype value https://github.com/mrdoob/three.js/pull/19997/commits/a53fcce912e08bb6dd7b85729ec9c5739420bd3e

**Comment by mrdoob** [comment]
Thanks!

**Comment by yomotsu** [comment]
Thank you so much for your kind advice and all help🙇‍♂️

**Comment by yomotsu** [comment]
FYI:
`this.isVector3` is not allowed because of the test: https://github.com/mrdoob/three.js/blob/e9e5d0e64bf154e17ba84a0f4d5ff0ea9003b436/test/unit/src/core/Object3D.tests.js#L243-L247
The properties of Vector3 can only contain `x`, `y`, `z`, and not `isVector3`.
But `isVector3.prototype.isVector3` can pass the test.

I think we should declare `is**` as `TheClass.prototype.is***`

cc @DefinitelyMaybe @ianpurvis

**Comment by DefinitelyMaybe** [comment]
thank you! I was starting to look into why my unit tests were failing and was approaching a similar conclusion. I hadn't figured out how to work around it though. Great stuff!

**Pull request merged**